proof/refine/ARM/Ipc_R.thy

```diff
@@ -5881,6 +5838,318 @@ lemma si_blk_makes_runnable':
+                            apply (frule (1) valid_objs_ko_at, clarsimp simp: valid_obj_def)
+                            apply (prop_tac "sc_relation sc n ko")
+                             apply (clarsimp simp: obj_at_def)
+                             apply (frule (1) pspace_relation_absD[OF _ state_relation_pspace_relation])
+                             apply (clarsimp simp: other_obj_relation_def obj_at'_def projectKOs )
+                            apply (frule (1) sc_ko_at_valid_objs_valid_sc', clarsimp)
+                            apply (subgoal_tac "0 < scRefillMax ko")
+                             apply (subgoal_tac "sc_valid_refills sc")
+                              apply (prop_tac "koa=ko", erule (1) ko_at'_inj, simp only:)
+                              apply (subgoal_tac "std_sc' ko")
+                               apply (simp add: sr_sc_refill_ready sr_sc_refill_sufficient state_relation_def active_sc_at_equiv)
+                              apply (metis std_sc'_normal sc_valid_refills_scRefillCount)
+                             apply (subst (asm) active_sc_at_equiv)
+                             apply (frule (1) active_sc_valid_refillsE)
+                             apply (clarsimp simp: valid_refills_def2 obj_at_def)
+                            apply (clarsimp simp: obj_at_def vs_all_heap_simps active_sc_def
+                                                  sc_relation_def)
+                           apply wpsimp
+                          apply (wpsimp simp: refillSufficient_def getRefills_def obj_at'_def)
+                         apply (wpsimp simp: refillReady_wp)
+                        apply (simp add: refillReady_def, rule no_ofail_gets_the)
+                        apply wpsimp
+                       apply wpsimp
+                      apply (clarsimp simp: obj_at_def sc_at_pred_n_def get_sched_context_exs_valid)
+                     apply (rule get_sched_context_exs_valid)
+                     apply (clarsimp simp: sc_at_pred_n_def obj_at_def)
+                    apply (wpsimp wp: get_sched_context_no_fail)
+                    apply (clarsimp simp: sc_at_pred_n_def obj_at_def is_sc_obj)
+                    apply (frule_tac sc=sc and n=n in valid_objs_valid_sched_context_size[OF invs_valid_objs], fastforce, simp)
+                   apply simp
+                  apply clarsimp
+                 apply (wpsimp wp: thread_get_wp')
+                apply (wpsimp wp: threadGet_wp)
+               apply (wpsimp wp: gts_wp)
+              apply (wpsimp wp: gts_wp')
+             apply (rule_tac Q="\<lambda>_. tcb_at recvr and valid_sched_action and invs and valid_list
+                                    and valid_release_q and scheduler_act_not recvr
+                                    and active_sc_valid_refills
+                                    and active_if_bound_sc_tcb_at recvr
+                                    and not_queued recvr"
+           in hoare_strengthen_post[rotated])
```

Comment:
nitpick: indentation (and there's `hoare_post_imp` if you don't want to have to rotate)